File: tiktok_bot.py
import requests
import json
import gspread
import os
from oauth2client.service_account import ServiceAccountCredentials
from datetime import datetime

# 1. Kết nối Google Sheet (Dùng chung Secret cũ của bạn)
scope = ["https://spreadsheets.google.com/feeds", "https://www.googleapis.com/auth/drive"]
# GitHub sẽ tự nạp file credentials.json từ Secret GOOGLE_CREDS bạn đã cài
creds = ServiceAccountCredentials.from_json_keyfile_name('credentials.json', scope)
client = gspread.authorize(creds)

# 2. Mở Sheet TikTok (Mai Thu nhớ tạo thêm 1 Tab tên là 'TikTok' trong file Google Sheet cũ nhé)
spreadsheet = client.open("Article") 
sheet = spreadsheet.worksheet("TikTok")

def crawl_tiktok():
    # API lấy danh sách bài viết mới nhất từ TikTok Academy
    url = "https://seller-vn.tiktok.com/university/api/knowledge/get_knowledge_list"
    params = {
        "page_size": 10,
        "page_number": 1,
        "region": "VN"
    }
    
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    }
    
    response = requests.get(url, params=params, headers=headers)
    data = response.json()
    
    if data.get('code') == 0:
        items = data['data']['list']
        existing_ids = sheet.col_values(1) # Lấy cột Knowledge_ID để check trùng
        
        current_date = datetime.now().strftime('%Y-%m-%d') # Định dạng yyyy-MM-dd cho PA dễ đọc
        
        for item in items:
            k_id = str(item['knowledge_id'])
            title = item['title']
            link = f"https://seller-vn.tiktok.com/university/essay?knowledge_id={k_id}"
            
            if k_id not in existing_ids:
                # Ghi vào Google Sheet: ID, Tiêu đề, Link, Ngày quét
                sheet.append_row([k_id, title, link, current_date])
                print(f"✅ Đã thêm TikTok: {title}")
            else:
                print(f"⏭️ Đã có: {title}")

if __name__ == "__main__":
    crawl_tiktok()

File: tiktok_bot.js
const { GoogleSpreadsheet } = require('google-spreadsheet');
const { JWT } = require('google-auth-library');
const axios = require('axios');

async function runTikTokBot() {
    const serviceAccountAuth = new JWT({
        email: JSON.parse(process.env.GOOGLE_SERVICE_ACCOUNT_JSON).client_email,
        key: JSON.parse(process.env.GOOGLE_SERVICE_ACCOUNT_JSON).private_key,
        scopes: ['https://www.googleapis.com/auth/spreadsheets'],
    });

    // NHỚ THAY ID SHEET CỦA BẠN VÀO ĐÂY NHA Mai Thu
    const doc = new GoogleSpreadsheet('1eAqPpi-ZyPEbTSDWw8OE1ngv07jjiwUAQy-XPYMutdY', serviceAccountAuth); 
    await doc.loadInfo();
    const sheet = doc.sheetsByTitle['TikTok'];

    const collections = [
        { id: "4227429090887426", chuDe: "Tài chính", danhMuc: "Phí của nhà bán hàng" }
    ];

    const now = new Date().toLocaleString('vi-VN', { timeZone: 'Asia/Ho_Chi_Minh' });
    const rows = await sheet.getRows();
    const existingIds = rows.map(row => row.get('Knowledge_ID'));

    for (const col of collections) {
        console.log(`--- Đang quét Content ID: ${col.id} ---`);
        
        // URL mới Mai Thu vừa tìm thấy nè
        const url = `https://seller-vn.tiktok.com/api/v1/seller/learning_center/contents/list`;
        
        try {
            const response = await axios.get(url, {
                params: {
                    content_id: col.id,
                    page: 1,
                    limit: 20,
                    region: "VN",
                    locale: "vi-VN"
                },
                headers: {
                    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
                    "Referer": "https://seller-vn.tiktok.com/university/home"
                }
            });

            // TikTok trả về data trong learning_info (như cái JSON bạn gửi lúc sáng)
            if (response.data && response.data.data && response.data.data.learning_info) {
                const items = response.data.data.learning_info;

                for (const item of items) {
                    const kId = item.learning_id.toString();
                    
                    const createDate = new Date(item.create_time).toLocaleDateString('vi-VN');
                    const modifyDate = new Date(item.modify_time).toLocaleDateString('vi-VN');

                    if (!existingIds.includes(kId)) {
                        await sheet.addRow({
                            'Knowledge_ID': kId,
                            'Tên bài': item.name,
                            'Ngày tạo': createDate,
                            'Ngày chỉnh sửa': modifyDate,
                            'Ngày giờ quét': now,
                            'Chủ đề': col.chuDe,
                            'Danh mục': col.danhMuc,
                            'Tóm tắt': item.description || "Không có tóm tắt",
                            'Link': `https://seller-vn.tiktok.com/university/essay?knowledge_id=${kId}`
                        });
                        console.log(`✅ Thêm bài: ${item.name}`);
                    } else {
                        console.log(`⏭️ Đã có: ${item.name}`);
                    }
                }
            } else {
                console.log(`⚠️ Không lấy được list bài. Kiểm tra lại API: ${JSON.stringify(response.data)}`);
            }
        } catch (error) {
            console.error(`❌ Lỗi API:`, error.message);
        }
    }
}

runTikTokBot().catch(console.error);
